### ISSUE TITLE
Negative-width bounding boxes when running on M1 (mps) HW

### ISSUE THREAD

**Issue by agentmorris** [ultralytics/yolov5#12645]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I am running a trained YOLOv5x6 model using val.py with the --save_json option.  I have a few images where the resulting .json file includes one or more boxes with negative width values (not negative x or y values, which seem normal and are discussed in other issues, but negative _width_ values), but I have only observed this behavior when running on M1 hardware (with the "mps" device).  All the boxes I've observed that have this property are low-ish confidence, but neither the confidence nor the absolute width/height are so low that I would call this a "junk bounding box".

I get identical, non-negative boxes processing the same images with the same weights on CUDA or CPU devices.  Just in case the default precision is different on M1, I tried this with and without half-precision inference, but could not replicate the negative boxes on CUDA/CPU devices.

So, this is a two-part question:

1. Are negative-width bounding boxes expected, and/or is there an interpretation of these boxes?  Should I expect to take the absolute value of the width and add it to the x value as if it were positive to get the right edge of the box, or should still add the (negative) width to the x value to get the other edge of the box?

2. In general, are there common patterns of differences between M1 and CUDA/CPU devices?  Are results expected to be identical to reasonable precision?

If the answers are "no, that should never happen" and "no, they should all be identical", I will try to provide a reproducible example.  I don't currently have permission to share the offending images, so I'm trying to see whether there's a logical explanation before going down that path.

Thanks!

-Dan

**Comment by glenn-jocher** [comment]
@agentmorris hi Dan,

Thanks for reaching out with your detailed observations. Negative-width bounding boxes are indeed not expected behavior, and there is no standard interpretation for such boxes in the context of YOLOv5. The bounding box coordinates should always result in positive width and height values, as they represent the dimensions of the detected object.

Regarding your two-part question:

1. Negative-width bounding boxes are not expected. The width (and height) should always be positive, as they represent the size of the bounding box. If you encounter negative values, it's likely indicative of an issue that needs to be addressed.

2. While results between M1 (mps) and CUDA/CPU devices should be very similar, minor differences can occur due to floating-point precision and hardware-specific optimizations. However, these differences should not result in negative dimensions for bounding boxes.

It seems like you might have stumbled upon a hardware-specific issue with the M1 (mps) device. If you can provide a reproducible example without sharing the original images, that would be very helpful. You might try to replicate the issue with publicly available images or synthetic data that triggers the same behavior.

In the meantime, I recommend checking the latest updates and documentation for any related fixes or advice on handling such cases. If the issue persists, please open an issue with as much detail as possible, and we'll look into it further.

**Comment by agentmorris** [comment]
Thank you for the quick response; I will try to create a reproducible example on a public image.

**Comment by glenn-jocher** [comment]
@agentmorris You're welcome, Dan! Creating a reproducible example on a public image would be extremely helpful. If you encounter any further issues or have additional questions while doing so, please don't hesitate to reach out. Good luck, and we appreciate your efforts in helping improve YOLOv5! 😊👍

**Comment by agentmorris** [comment]
I am able to share images that reproduce this behavior now, and I also have some new data that might get us closer to a root cause:

- In a Python environment based on the _current_ YOLOv5 repo, I am able to reproduce this negative-box behavior on an M1 Mac (specifically a mac2.metal M1 AWS VM).
- In a Python environment with older versions of most dependencies, with an old version of the YOLOv5 repo (based on [this commit](https://github.com/ultralytics/yolov5/commit/c23a441c9df7ca9b1f275e8c8719c949269160d1/)), on the same M1 machine, I am *not* able to reproduce this negative-box behavior.
- Either Python/YOLOv5 environment works fine (i.e., does not produce negative-width boxes) on a Linux+CUDA or Linux+CPU environment.
- The Linux+CUDA results in either environment are nearly identical to the "old YOLOv5" M1 results, with just acceptable differences out at the fourth or fifth decimal place.

The command I am running on the M1 VM is:

`python val.py --task test --data "/Users/ec2-user/images/dataset.yaml" --weights "/Users/ec2-user/md_v5a.0.0.pt" --batch-size 1 --imgsz 1280 --conf-thres 0.001 --device "mps" --save-json --project "/Users/ec2-user/yolo-results/yolo-new-no-aug" --name "yolo_results" --exist-ok --save-txt --save-conf
`

- Images are [here](http://dmorris.net/misc/tmp/m1-yolo-issue/m1-test-images.zip) (images.zip, three small images)
- Model weights are [here](https://github.com/agentmorris/MegaDetector/releases/download/v5.0/md_v5a.0.0.pt) (md_v5a.0.0.pt, a trained YOLOv5 model)
- A suitable dataset.yaml file is [here](http://dmorris.net/misc/tmp/m1-yolo-issue/dataset.yaml) (the absolute path on the first line will need adjusting)
- The COCO-formatted results from the "current YOLOv5" environment are [here](http://dmorris.net/misc/tmp/m1-yolo-issue/yolo-new-md_v5a.0.0_predictions.json) (with negative-width boxes)
- The COCO-formatted results from the "old YOLOv5 "environment are [here](http://dmorris.net/misc/tmp/m1-yolo-issue/yolo-old-md_v5a.0.0_predictions.json) (no negative boxes, nearly identical to the results from a CUDA environment)
- pip's list of packages in the "current YOLOv5" environment is [here](http://dmorris.net/misc/tmp/m1-yolo-issue/yolov5-new-negative-width-pip.txt).  This directly follows "pip install -r requirements.txt"; I made no other changes to this environment.
- pip's list of packages in the "old YOLOv5" environment is [here](http://dmorris.net/misc/tmp/m1-yolo-issue/yolov5-old-working-pip.txt).  This is not strictly a pip install of the requirements.txt file from the (old) YOLOv5 repo, but in this case, I don't think it matters for purposes of reproducing the issue, since this is the environment that *doesn't* exhibit the behavior in question.

--

Thanks for any ideas you have!  Hopefully there is user error in here somewhere.

**Comment by glenn-jocher** [comment]
@agentmorris Thank you for providing a detailed follow-up with the resources to reproduce the issue. It's great that you've narrowed down the circumstances under which the negative-width bounding boxes occur and that you've identified a difference in behavior between the current and older YOLOv5 versions on M1 hardware.

Given the information you've provided, it seems like there may have been changes introduced after the older commit that could be affecting the bounding box calculations on M1 (mps) devices specifically. The next steps would be to:

1. Investigate the differences in the codebase between the old commit and the current version that could potentially affect bounding box calculations on M1 hardware.
2. Run a bisect between the two versions to identify the specific commit that introduced the negative-width issue.
3. Examine the changes in dependencies that could have an impact on the results.

Since you've provided the necessary files and environment details, I or a member of the Ultralytics team will look into this issue further. We aim to ensure consistency across different hardware platforms, and your detailed report is instrumental in helping us achieve that.

Please feel free to open an issue on the YOLOv5 GitHub repository with the information you've shared here, so we can track the progress on this matter and involve more contributors if needed.

Your efforts in identifying and documenting this issue are much appreciated! 🙏

**Comment by agentmorris** [comment]
> Please feel free to open an issue on the YOLOv5 GitHub repository with the information you've shared here, so we can track the progress on this matter and involve more contributors if needed.

I created [issue 12654](https://github.com/ultralytics/yolov5/issues/12654).  Thanks for taking a look!

**Comment by glenn-jocher** [comment]
@agentmorris Thank you for creating the issue on the YOLOv5 GitHub repository. We'll take it from there and investigate the problem further. Your detailed report and the resources you've provided will be very helpful in diagnosing and hopefully resolving the issue.

Keep an eye on the issue tracker for updates, and feel free to contribute any additional information or insights you may have as we work through this. Your collaboration is invaluable, and we're committed to improving YOLOv5 for all users and platforms.

Thanks again for your contribution! 👍😊

**Comment by agentmorris** [comment]
Hmmm... the GHA bot did in fact close this issue, in contrast to what @glenn-jocher suggested a couple days ago.  Is this really not planned, or is that an overly-aggressive GHA bot?

**Comment by guybashan** [comment]
I am getting the exact same behavior. Many if my images are with negative width. I am with M1 max.

**Comment by pderrenger** [comment]
Please ensure you're using the latest version of YOLOv5. If the issue persists, consider sharing a minimal reproducible example on the GitHub issue tracker for further investigation.

**Comment by guybashan** [comment]
> Hi, please ensure you're using the latest version of YOLOv5. If the issue persists, let us know with any additional details.

Hi,

- Indeed I am using the latest code from git (main branch).
- I re-run the training using this:
`/cam-analyzer/pytorch_metal_env/bin/python ../yolov5/train.py --img 416 --batch 16 --epochs 100 --data ../yolo_cfg/data.yaml --weights yolov5m.pt --device mps`
- I re-checked my "train" and "val" folders. They look ok and the files look fine. I created the labels using YoloLabel app (open source).
- I run the detection using this command: 
`/cam-analyzer/pytorch_metal_env/bin/python detect.py --save-txt --save-conf --weights /Users/user/Documents/dev/yolo_model/cam_analyzer/weights/best.pt --source /Users/user/Documents/dev/yolo_images --conf 0.1 --project /Users/columbo/Documents/dev/yolo_results --name detections --device mps`
- I use M1 Max with 64GB memory
- I enabled working with GPU

**Comment by pderrenger** [comment]
@guybashan thank you for the details. Since you're using the latest version and still experiencing issues, please check if the problem occurs with different images or datasets. If it persists, consider opening a GitHub issue with your findings for further investigation.

**Comment by guybashan** [comment]
Indeed, it happens with different images. As I already wrote:

I changed from "mps" to "cpu" and now the **weight** logs with no negative values.
The downside: instead of avg detection for ~10ms I went down to: ~70ms.

One more important thing: I didn't rerun the creation of the model. So I can only guess the issue is in the "detect.py" and not in the "train.py".

**Comment by pderrenger** [comment]
Thank you for the update. It seems like the issue might be specific to the "mps" device during detection. Please ensure your environment and dependencies are up to date. If the problem continues, consider opening a GitHub issue with your findings for further investigation.

**Comment by csbrown** [comment]
Is there anything constraining the output to be positive?  I have a use case where I would like to have possibly negative w/h for my bounding boxes.  I am wondering if there is any special branching code that actively attempts to prevent this.

**Comment by pderrenger** [comment]
The YOLOv5 architecture inherently predicts positive widths and heights for bounding boxes due to its design, which uses exponential functions to ensure these values remain positive. There is no special branching code to enforce this; it is a result of the mathematical formulation of the model. If you require negative dimensions, you would need to modify the model's bounding box prediction logic, though this is not standard and could introduce instability. For details, you can reference YOLOv5's [bounding box formulation](https://docs.ultralytics.com/yolov5/tutorials/architecture_description/#43-eliminate-grid-sensitivity).
